### DISCUSSION TITLE
model.transcribe() modified to perform batch inference on audio files

### DISCUSSION THREAD

**Discussion by Blair-Johnson** [openai/whisper#662] [Show and tell]
**What's Different**
A few people have posted questions asking about how whisper could be used to efficiently process audio clips in parallel. The underlying encoder and decoder both support batched inference, so it should be possible to batch clips together for increased throughput on GPUs. I modified the implementation of the `transcribe()` function to branch to an alternate `batch_transcribe()` whenever the user supplies a list of audio files. Everything is essentially the same as the default `model.transcribe()` implementation, however the mel spectrograms and conditional prompts are batched together for the model inference stages of the pipeline. This allows us to get substantially sub-linear scaling of throughput on GPUs that have additional head-room when running serial transcription.

**Example usage:**
```python
import whisper

model = whisper.load_model("base")
results = model.transcribe(["audio1.mp3", "audio2.mp3"])
print(results[0]['text'])
print(results[1]['text'])
```
**Results**
Testing transcription on a 3.5 hour podcast batched together with itself in groups of 1, 2, 4, 8, 16, and 32 we can see that we get significant speedups through batching on a NVIDIA A100 (this is the `largev1` model). We see sub-linear scaling until a batch size of 16, after which the GPU becomes saturated and the scaling becomes linear (but still 3-5x higher throughput than serial).

<img width="1028" alt="image" src="https://user-images.githubusercontent.com/43356315/206748537-8ee7fa90-277e-4fc4-806d-4533ae9ff43d.png">

When clips of different lengths are used, the internal model batch sizes will be reduced whenever a shorter clip is done being transcribed. This means that it is more efficient to batch transcribe clips of a similar length together. A possible pipeline for many audio clips would involve sorting them by length and batching neighboring clips by some optimal batch size for the GPU in question.

You can check out the batched version of whisper in my fork [here](https://github.com/Blair-Johnson/batch-whisper/tree/main).

I haven't tested every use-case to verify that these modifications are 100% non-breaking so there's no PR at the moment.

**Comment by frankiedrake** [comment]
How does it work for you? I tried to test it but it seems it doesn't work, I'm getting an error 
```
File "/home/.../python3.8/site-packages/whisper/decoding.py", line 622, in run
    tokens: Tensor = torch.tensor([self.initial_tokens]).repeat(n_audio, 1)
TypeError: must be real number, not list
```

**Reply by Blair-Johnson** [reply]
That line + line number combination seems to match the main repository. These changes haven't been merged into the main project; try installing from the forked repo that I [linked](https://github.com/Blair-Johnson/batch-whisper/tree/main). Let me know if you still have issues!

**Reply by frankiedrake** [reply]
> That line + line number combination seems to match the main repository. These changes haven't been merged into the main project; try installing from the forked repo that I [linked](https://github.com/Blair-Johnson/batch-whisper/tree/main). Let me know if you still have issues!

But I copied your `transcribe.py` file and use `batch_transcribe` method from there, does what you say still relevant?

**Reply by frankiedrake** [reply]
I installed library from your repo, but still can't get it working(
Now I get
```
Traceback (most recent call last):
  File "/home/.../python3.8/site-packages/whisper/transcribe.py", line 477, in batch_transcribe
    batch_timestamp_tokens: List[torch.Tensor] = [tokens.ge(tokenizers[languages[imap[i]]].timestamp_begin) for i,tokens in enumerate(batch_tokens)]
  File "/home/.../python3.8/site-packages/whisper/transcribe.py", line 477, in <listcomp>
    batch_timestamp_tokens: List[torch.Tensor] = [tokens.ge(tokenizers[languages[imap[i]]].timestamp_begin) for i,tokens in enumerate(batch_tokens)]
IndexError: list index out of range
```

**Reply by Blair-Johnson** [reply]
Yes, there are other changes outside of `transcribe.py` that are necessary to support the `batch_transcribe` method. You can take a look at the changes in the most recent [PR](https://github.com/Blair-Johnson/batch-whisper/commit/fb6159b43ab5b6a47a4d73d7420d8f7c531d5144) on my fork to see what exactly was changed.

**Reply by Blair-Johnson** [reply]
Could you provide an example of your usage to reproduce the error?

**Reply by frankiedrake** [reply]
I'm using it this way:
`transcribe(model, audio=files, language='uk', fp16=False)`

**Reply by frankiedrake** [reply]
I got it... I passed `language` parameter, and it was breaking because you define `languages = []` at line 337 and then, if language is passed in `options` parameter - this value isn't updating. 
Maybe we can make it possible to provide lang for each file? I have some logic on it so I can't totally rely on whisper's lang detection

**Reply by Blair-Johnson** [reply]
It's definitely the language handling. I've tested with auto-detecting language for each clip, but things may break if you try to broadcast a single language to all the clips like that. It should be an easy-enough fix though.

**Reply by frankiedrake** [reply]
Should I deal with it myself or I can make a fork with such functionality?

**Reply by Blair-Johnson** [reply]
I've fixed the issue and added the ability to broadcast languages and specify a list of languages. I'm going to do some tests then update the repo sometime today.

**Reply by Blair-Johnson** [reply]
@frankiedrake The repo has been updated. You should now be able to specify language as a list on a per-clip basis: `model.transcribe(['clip1.mp3', 'clip2.mp3'], language=['en', 'uk'])` or with a single string broadcast to all clips `model.transcribe(['clip1.mp3', 'clip2.mp3'], language='uk')`. Let me know if you have additional issues!

**Reply by frankiedrake** [reply]
I've just fixed probably a typo in `decoding.py:77`
`language: Optional[Union[str, List[str]]] = None` 
instead of 
`language: Optional[Union[str, list[str]]] = None` 
and it seems working, thank you!

**Reply by frankiedrake** [reply]
Hello, again @Blair-Johnson, Just noticed that passing language param as string or omitting it causes errors in `decoding.py:620`
    `return [tuple(tokens) for tokens in res_tokens]
TypeError: 'int' object is not iterable`

I tried to fix it but it causes errors in other places, so I decided not to mess with it)

**Reply by Blair-Johnson** [reply]
Could you provide an example to reproduce the issue? You should also be able to file an issue on the repo- that'll give you more space to share details.

**Reply by frankiedrake** [reply]
Sorry, I guess I messed up with installation repositories. I think it is okay now

**Comment by kinghmy** [comment]
你好，来信我已收到，我会尽快处理，谢谢！

**Comment by qqfly1to19** [comment]
results = model.transcribe(["audio1.mp3", "audio2.mp3"], ,initial_prompt=prompt) the prompt dose not work.  see. https://github.com/openai/whisper/discussions/277
```
prompt='以下是普通话的句子'
result = model.transcribe([audioFiles],language='zh',verbose=True,initial_prompt=prompt)
```

**Comment by huynhthanh98** [comment]
Hi @Blair-Johnson !

Thank you very much. Your code runned quickly but when i used beam_size=5 the code returned the error

`ValueError: torch.Size([1, 3])[0] % 5 != 0`

Can you help me fix this error?

**Reply by Blair-Johnson** [reply]
The beam search decoder isn't supported at the moment, and this is an error related to that. I have an experimental version where the beam search options work, but the changes introduced performance issues in the default decoding behavior. It needs more work; I can create a [branch](https://github.com/Blair-Johnson/batch-whisper/tree/beam_search_experimental) with those changes for you.

**Comment by a-froghyar** [comment]
Hey @Blair-Johnson, thanks for your great contribution! When I try to run batch processing using your fork (using batch size of 16), I'm experiencing significantly lower inference speed compared to vanilla inference (it takes longer using batch processing than with single example inference). I'm trying to transcribe ~10hrs of data segmented into ~50s segments. Do you have any pointers? Thanks!

**Reply by Blair-Johnson** [reply]
I would try tuning the batch size that you use to find an optimal value for your GPU. Even on an A100, the optimal batch size is probably <16 for the large model. Also, just to be clear, it's expected for a batch to take longer than a serial inference. If you're using a batch size of 16 and inference on a batch takes less than 16x longer than serial processing, then that represents a speedup. I don't think I've experienced a case where the batched version is slower than running serially, so please open an issue and provide more details for reproducing the behavior if you think there's a bug, I'd be happy to help.

**Reply by chesha1** [reply]
> Hey @Blair-Johnson, thanks for your great contribution! When I try to run batch processing using your fork (using batch size of 16), I'm experiencing significantly lower inference speed compared to vanilla inference (it takes longer using batch processing than with single example inference). I'm trying to transcribe ~10hrs of data segmented into ~50s segments. Do you have any pointers? Thanks!

If you would like to cut long audio into shorter slice, I recommend this project:
https://pytorch.org/hub/snakers4_silero-vad_vad/
Actually, Whisper is trained on audios of 30s or shorter, so it does not work well on long audio.

**Reply by Blair-Johnson** [reply]
Whisper does work well on long audio, but under the hood it breaks it into 30s segments and processes them separately. It will however, condition the text it generates for a segment on the predicted text from the previous segment to improve continuity. If you break a longer clip into segments and process them in batches non-sequentially, then you'll want to disable this conditioning. The quality of transcription will likely suffer. Audio batching works better when you have a number of different clips you'd like processed together.

**Reply by chesha1** [reply]
> Whisper does work well on long audio, but under the hood it breaks it into 30s segments and processes them separately. It will however, condition the text it generates for a segment on the predicted text from the previous segment to improve continuity. If you break a longer clip into segments and process them in batches non-sequentially, then you'll want to disable this conditioning. The quality of transcription will likely suffer. Audio batching works better when you have a number of different clips you'd like processed together.

Thanks for your explaination. I have another question: when whisper processes audio which is shorter than 30s, it will also give an output of segmentation. Is this segmentation the same thing you mentioned above? The long audio is segmented by semantic information of text content or audio feature?

**Reply by Blair-Johnson** [reply]
> > Whisper does work well on long audio, but under the hood it breaks it into 30s segments and processes them separately. It will however, condition the text it generates for a segment on the predicted text from the previous segment to improve continuity. If you break a longer clip into segments and process them in batches non-sequentially, then you'll want to disable this conditioning. The quality of transcription will likely suffer. Audio batching works better when you have a number of different clips you'd like processed together.
> 
> Thanks for your explaination. I have another question: when whisper processes audio which is shorter than 30s, it will also give an output of segmentation. Is this segmentation the same thing you mentioned above? The long audio is segmented by semantic information of text content or audio feature?

Could you clarify what you mean by segmentation? I believe that clips shorter than 30s get padded to 30s before they're processed.

**Reply by chesha1** [reply]
> > > Whisper does work well on long audio, but under the hood it breaks it into 30s segments and processes them separately. It will however, condition the text it generates for a segment on the predicted text from the previous segment to improve continuity. If you break a longer clip into segments and process them in batches non-sequentially, then you'll want to disable this conditioning. The quality of transcription will likely suffer. Audio batching works better when you have a number of different clips you'd like processed together.
> > 
> > 
> > Thanks for your explaination. I have another question: when whisper processes audio which is shorter than 30s, it will also give an output of segmentation. Is this segmentation the same thing you mentioned above? The long audio is segmented by semantic information of text content or audio feature?
> 
> Could you clarify what you mean by segmentation? I believe that clips shorter than 30s get padded to 30s before they're processed.

model.transcribe() returns a dict. This dict has an element whose key is 'segments' and value is a list containing text, timestamps and many other parameters. Sometimes length of this list is more than 1, even when input audio is shorter than 30s.

**Reply by Blair-Johnson** [reply]
> > > > Whisper does work well on long audio, but under the hood it breaks it into 30s segments and processes them separately. It will however, condition the text it generates for a segment on the predicted text from the previous segment to improve continuity. If you break a longer clip into segments and process them in batches non-sequentially, then you'll want to disable this conditioning. The quality of transcription will likely suffer. Audio batching works better when you have a number of different clips you'd like processed together.
> > > 
> > > 
> > > Thanks for your explaination. I have another question: when whisper processes audio which is shorter than 30s, it will also give an output of segmentation. Is this segmentation the same thing you mentioned above? The long audio is segmented by semantic information of text content or audio feature?
> > 
> > 
> > Could you clarify what you mean by segmentation? I believe that clips shorter than 30s get padded to 30s before they're processed.
> 
> model.transcribe() returns a dict. This dict has an element whose key is 'segments' and value is a list containing text, timestamps and many other parameters. Sometimes length of this list is more than 1, even when input audio is shorter than 30s.

Ah, yes. The spectrograms that the model ingests represent 30s of audio, but the 'segments' that the model returns are of variable length and determined by the transformer output. Essentially a segment is the series tokens that the model outputs between start and end timestamp tokens. So when the model generates several timestamped sequences of text for a 30s audio clip you will get several returned segments containing information about the transcription at different times in the 30s clip (ex: 0-5s, 5-8s, 8-15s, etc.).

**Comment by jake1271** [comment]
This is incredible!
Regarding the commandline usage, when specifying multiple file, will they also be processed in batch mode?

**Reply by Blair-Johnson** [reply]
At the moment, no. It would be fairly easy to make that modification though; you would just need to pass the filenames in as a list to the transcribe method [here](https://github.com/Blair-Johnson/batch-whisper/blob/76b3f818fb94a7c18c37d2ffecf71d7c69182749/whisper/transcribe.py#L609). Result will then be a list of transcripts rather than a single transcript, so you will need to iterate over it to write each one to a file.

**Comment by treya-lin** [comment]
To use this repo I need to create a new environment and instead of installing the official whisper I should install the whisper from this repo right?

**Reply by Blair-Johnson** [reply]
Yes, that's correct.

**Comment by S1E9** [comment]
Hi there! OpenAI just release its latest large-v3 model in whisper, is it possiable for batch-whisper to support this update?

**Reply by Blair-Johnson** [reply]
I would take a look at the huggingface pipelines for whisper-v3, I think they support batch decoding. https://huggingface.co/openai/whisper-large-v3

**Comment by kinghmy** [comment]
你好，来信我已收到，我会尽快处理，谢谢！

**Comment by vadimkantorov** [comment]
@Blair-Johnson Maybe you should open a draft PR anyway, it'd be a bit easier to inspect the diff and so on :)

**Comment by kinghmy** [comment]
你好，来信我已收到，我会尽快处理，谢谢！

**Comment by mohith7548** [comment]
I get ValueError with this implementation:

```
ValueError: tuple.index(x): x not in tuple
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<timed exec> in <module>

/local_disk0/.ephemeral_nfs/envs/pythonEnv-e4288798-fca8-417f-8b49-f0a77d8c8ba0/lib/python3.8/site-packages/whisper/transcribe.py in transcribe(model, audio, verbose, temperature, compression_ratio_threshold, logprob_threshold, no_speech_threshold, condition_on_previous_text, **decode_options)
     73     """
     74     if type(audio) == list:
---> 75         return batch_transcribe(model=model,
     76                                 audio=audio,
     77                                 verbose=verbose,

/local_disk0/.ephemeral_nfs/envs/pythonEnv-e4288798-fca8-417f-8b49-f0a77d8c8ba0/lib/python3.8/site-packages/whisper/transcribe.py in batch_transcribe(model, audio, verbose, temperature, compression_ratio_threshold, logprob_threshold, no_speech_threshold, condition_on_previous_text, **decode_options)
    469             decode_options["prompt"] = [all_tokens[imap[i]][prompt_reset_since[imap[i]]:] for i in range(len(batch_segments))]
    470             decode_options["language"] = [l for i,l in enumerate(languages) if continue_processing[i]]
--> 471             results: List[DecodingResult] = decode_with_fallback(torch.stack(batch_segments))
    472             batch_tokens = [torch.tensor(result.tokens) for result in results]
    473 

/local_disk0/.ephemeral_nfs/envs/pythonEnv-e4288798-fca8-417f-8b49-f0a77d8c8ba0/lib/python3.8/site-packages/whisper/transcribe.py in decode_with_fallback(segment)
    377 
    378             options = DecodingOptions(**kwargs, temperature=t)
--> 379             decode_result = model.decode(segment, options)
    380 
    381             needs_fallback = False

/databricks/python/lib/python3.8/site-packages/torch/utils/_contextlib.py in decorate_context(*args, **kwargs)
    113     def decorate_context(*args, **kwargs):
    114         with ctx_factory():
--> 115             return func(*args, **kwargs)
    116 
    117     return decorate_context

/local_disk0/.ephemeral_nfs/envs/pythonEnv-e4288798-fca8-417f-8b49-f0a77d8c8ba0/lib/python3.8/site-packages/whisper/decoding.py in decode(model, mel, options)
    858         mel = mel.unsqueeze(0)
    859 
--> 860     result = DecodingTask(model, options).run(mel)
    861 
    862     if single:

/local_disk0/.ephemeral_nfs/envs/pythonEnv-e4288798-fca8-417f-8b49-f0a77d8c8ba0/lib/python3.8/site-packages/whisper/decoding.py in __init__(self, model, options)
    481             # branch to handle batched case
    482             self.sample_begin: List[int] = [len(tokens) for tokens in self.initial_tokens]
--> 483             self.sot_index: List[int] = [tokens.index(self.tokenizers[i].sot) for i,tokens in enumerate(self.initial_tokens)]
    484 
    485             # inference: implements the forward pass through the decoder, including kv caching

/local_disk0/.ephemeral_nfs/envs/pythonEnv-e4288798-fca8-417f-8b49-f0a77d8c8ba0/lib/python3.8/site-packages/whisper/decoding.py in <listcomp>(.0)
    481             # branch to handle batched case
    482             self.sample_begin: List[int] = [len(tokens) for tokens in self.initial_tokens]
--> 483             self.sot_index: List[int] = [tokens.index(self.tokenizers[i].sot) for i,tokens in enumerate(self.initial_tokens)]
    484 
    485             # inference: implements the forward pass through the decoder, including kv caching

ValueError: tuple.index(x): x not in tuple
```

**Comment by kinghmy** [comment]
你好，来信我已收到，我会尽快处理，谢谢！

**Comment by patrickvonplaten** [comment]
BTW, we have just added support for batched long-form transcription to Transformers: https://github.com/huggingface/transformers/pull/27658. It will be in today's release.

With batched transcription and a batch size of 16, Transformers generate is 4x faster than this codebase.

Check the usage section in the PR description to try it out https://github.com/huggingface/transformers/pull/27658

**Reply by Blair-Johnson** [reply]
@mohith7548 I recommend using the transformers version.

**Reply by sidharthanup** [reply]
@patrickvonplaten I do not see info on how to process multiple mp3 files in parallel in the usage section of the PR. Can you help me with this?

**Reply by sidharthanup** [reply]
Never mind. I got it to work! This is what I did:

```python
import torchaudio 
from transformers import WhisperForConditionalGeneration, AutoProcessor
from datasets import load_dataset, Audio
import numpy as np
import torch
import time
import whisper

# Assuming files is the list of audio files
files = ["audio1.mp3", "audio2.mp3", "audio3.mp3", "audio4.mp3", "audio5.mp3"]

ds = load_dataset("/path/to/files/", data_files=files)["train"]
ds = ds.cast_column("audio", Audio(sampling_rate=16000))

raw_audio = [x["array"].astype(np.float32) for x in ds["audio"]]

processor = AutoProcessor.from_pretrained("openai/whisper-medium.en")
inputs = processor(raw_audio, return_tensors="pt", truncation=False, padding="longest", return_attention_mask=True, sampling_rate=16_000)
inputs = inputs.to("cuda", torch.float16)

model_medium = WhisperForConditionalGeneration.from_pretrained("openai/whisper-medium.en", torch_dtype=torch.float16)
model_medium.to("cuda")

# activate `temperature_fallback` and repetition detection filters and condition on prev text
result = model_medium.generate(**inputs, condition_on_prev_tokens=False, temperature=(0.0, 0.2, 0.4, 0.6, 0.8, 1.0), logprob_threshold=-1.0, compression_ratio_threshold=1.35, return_timestamps=True)

decoded = processor.batch_decode(result, skip_special_tokens=True)

```

**Comment by afsara-ben** [comment]
it doesnt work for me, i get error `TypeError: expected np.ndarray (got list)`

**Reply by AntyRia** [reply]
me too, do u fixed it?

**Reply by matlafu** [reply]
I might be wrong, but the way I understood it is that the batch functionality was available only in his fork of the repo. But that version has been deprecated, so it might not work anymore :( I encountered the same error today and could not make it work no matter what I tried

**Comment by sidharthanup** [comment]
In case people didn't see my comment, here's what I did to get it to work for a list of audio files : 

```python
import torchaudio 
from transformers import WhisperForConditionalGeneration, AutoProcessor
from datasets import load_dataset, Audio
import numpy as np
import torch
import time
import whisper

# Assuming files is the list of audio files
files = ["audio1.mp3", "audio2.mp3", "audio3.mp3", "audio4.mp3", "audio5.mp3"]

ds = load_dataset("/path/to/files/", data_files=files)["train"]
ds = ds.cast_column("audio", Audio(sampling_rate=16000))

raw_audio = [x["array"].astype(np.float32) for x in ds["audio"]]

processor = AutoProcessor.from_pretrained("openai/whisper-medium.en")
inputs = processor(raw_audio, return_tensors="pt", truncation=False, padding="longest", return_attention_mask=True, sampling_rate=16_000)
inputs = inputs.to("cuda", torch.float16)

model_medium = WhisperForConditionalGeneration.from_pretrained("openai/whisper-medium.en", torch_dtype=torch.float16)
model_medium.to("cuda")

# activate `temperature_fallback` and repetition detection filters and condition on prev text
result = model_medium.generate(**inputs, condition_on_prev_tokens=False, temperature=(0.0, 0.2, 0.4, 0.6, 0.8, 1.0), logprob_threshold=-1.0, compression_ratio_threshold=1.35, return_timestamps=True)

decoded = processor.batch_decode(result, skip_special_tokens=True)
```

**Comment by manish-kumar-iisc** [comment]
I am not able to use the below code, it's throwing an error for me
```
model = whisper.load_model("base")
results = model.transcribe(["audio1.mp3", "audio2.mp3"])

```
how to use that?

1. I have installed the whisper using [url](https://pypi.org/project/openai-whisper/).
2. Do i need to install from the [fork](https://github.com/Blair-Johnson/batch-whisper/tree/main)?
@Blair-Johnson 

Getting below error
![image](https://github.com/user-attachments/assets/c1038be7-2271-49d1-b57f-8f6fc0b5cf7c)

**Reply by Blair-Johnson** [reply]
You would need to install from the fork for the provided code snippets to work. There are a few bugs you may or may not encounter, and the code hasn't been maintained for newer versions of whisper. I would recommend taking a look at the [transformers implementation](https://huggingface.co/docs/transformers/en/model_doc/whisper) which should be up-to-date and have optimizations like flash attention and can use torch.compile. Even operating serially, you might see better overall throughput with these optimizations. Good luck!

**Reply by rkiotech** [reply]
I've made a fork that offers batch inference of audio files.
https://github.com/rkiotech/whisper_parallel_inferencing

**Reply by j-silv** [reply]
> I've made a fork that offers batch inference of audio files. https://github.com/rkiotech/whisper_parallel_inferencing

Thank you so much :) 

The author suggests a really straightforward method which simply stacks padded tensors and passes them directly to `whisper.decode`. Doesn't require any internal library modification, since the Whisper inference functions expect a batch dimension already. I'm surprised this isn't at least mentioned in the official README. 

I tested this method with 48 audio files ranging from 2 to 30 seconds, using large model, and I got a 3x time improvement.

**Comment by kinghmy** [comment]
你好，来信我已收到，我会尽快处理，谢谢！
